### ISSUE TITLE
GQL endpoint should accept batched HTTP requests

### ISSUE THREAD

**Issue by JeroenStravers1** [creativesoftwarefdn/weaviate#498]
Weaviate's graphql POST [receiver](https://github.com/creativesoftwarefdn/weaviate/blob/develop/restapi/operations/graphql/weaviate_graphql_post.go) does not currently accept batched requests, which is something we will need. Incoming network traffic could very well be batched. 

I'll examine if this is also true for regular HTTP requests.

cc: @dljcollette, @bobvanluijt 
related: https://github.com/SeMI-network/weaviate-graphql-prototype/issues/34

**Comment by JeroenStravers1** [comment]
Based on the internals of Weaviate's [HTTP Things Update handler](https://github.com/creativesoftwarefdn/weaviate/blob/1d4b50f8f2a9fb47615c0c29da6540de1f386a0e/restapi/operations/things/weaviate_things_update.go#L56) I'd say we haven't implemented this for regular HTTP Requests either. 

If I'm reading this correctly the update handler [instantiates](https://github.com/creativesoftwarefdn/weaviate/blob/1d4b50f8f2a9fb47615c0c29da6540de1f386a0e/restapi/operations/things/weaviate_things_update.go#L61) a single [WeaviateUpdateParams](https://github.com/creativesoftwarefdn/weaviate/blob/1d4b50f8f2a9fb47615c0c29da6540de1f386a0e/restapi/operations/things/weaviate_things_update_parameters.go#L45) struct, which can only contain a single HTTPRequest. The `WeaviateUpdateParams` is [passed](https://github.com/creativesoftwarefdn/weaviate/blob/1d4b50f8f2a9fb47615c0c29da6540de1f386a0e/restapi/operations/things/weaviate_things_update.go#L81) to the [Handle](https://github.com/creativesoftwarefdn/weaviate/blob/1d4b50f8f2a9fb47615c0c29da6540de1f386a0e/restapi/operations/things/weaviate_things_update.go#L30) function of the [WeaviateThingsUpdateHandlerFunc](https://github.com/creativesoftwarefdn/weaviate/blob/1d4b50f8f2a9fb47615c0c29da6540de1f386a0e/restapi/operations/things/weaviate_things_update.go#L27), at which point it ends up in `configure_weaviate.go`'s [update request handler](https://github.com/creativesoftwarefdn/weaviate/blob/1d4b50f8f2a9fb47615c0c29da6540de1f386a0e/restapi/configure_weaviate.go#L1053). This update handler does not include logic for iterating over a list of requests (nested or otherwise).

Interpreting an Apollo batched request, however, would require iteration:
```
[
  {
    query: < query 0 >,
    variables: < variables for query 0 >,
  },
  {
    query: < query 1 >,
    variables: < variables for query 1 >,
  },
  {
    query: < query n >
    variables: < variables for query n >,
  }
]
```

query example retrieved from the bottom of this [blogpost](https://blog.apollographql.com/query-batching-in-apollo-63acfd859862). Could you confirm the part about the Apollo batched request @dljcollette?

**Comment by dljcollette** [comment]
> query example retrieved from the bottom of this [blogpost](https://blog.apollographql.com/query-batching-in-apollo-63acfd859862). Could you confirm the part about the Apollo batched request @dljcollette?

Yes that's correct. Also see the [official docs](https://github.com/apollographql/apollo-server/blob/master/docs/source/requests.md).

**Comment by dljcollette** [comment]
You might also want to look at [this post](https://medium.com/safetycultureengineering/how-to-handle-graphql-query-batching-in-golang-3a16e6c3ac9c), though not very detailed it provides a bit more insight in the workings of query batching on the server.

**Comment by bobvanluijt** [comment]
Based on [this issue](https://github.com/creativesoftwarefdn/weaviate/issues/506) a suggestion for fast implementation.

1. Check [here](https://github.com/creativesoftwarefdn/weaviate/blob/develop/restapi/configure_weaviate.go#L1207) if you are receiving an array (slice) or an object (with `query.(type)`).
2. If it is _not_ a slice, continue as before.
3. If it is a slice (thus an array) loop through all the responses in a goroutine. You collect all the responses and return the result.

**Comment by moretea** [comment]
Hm. I don't think that you can do this, because it's already been casted to a concrete type. 

What might work is if you'd  make the body a `union` of a GraphQLRequest and a GraphQLRequests objects.

**Comment by bobvanluijt** [comment]
Ha yes, but the same principle applies. The request will be handled in `api.GraphqlWeaviateGraphqlPostHandler`.

**Comment by JeroenStravers1** [comment]
I agree with @moretea on this one. Lets take the Thing validation [endpoint](https://github.com/creativesoftwarefdn/weaviate/blob/e971fb87326f3b0eb37b7fd094b43153f927d0ca/restapi/configure_weaviate.go#L1115) as an example. It takes a [WeaviateThingsValidateParams](https://github.com/creativesoftwarefdn/weaviate/blob/e971fb87326f3b0eb37b7fd094b43153f927d0ca/restapi/operations/things/weaviate_things_validate_parameters.go#L42) struct as function input, and this struct can only contain a single HTTP-request. Feeding it an array would result in an error before the request even begins to be interpreted. I chose this endpoint randomly, the same would hold true for all existing endpoints (we are talking about all endpoints, right, not just the GQL one? If not, this applies to the GQL endpoint too ;) ). 

I am interested in this `union`, could you elaborate?

**Comment by bobvanluijt** [comment]
I'll setup a quick example...

**Comment by bobvanluijt** [comment]
@JeroenStravers1 I've created an example, please note: _IT IS PSEUDO CODE_ it is written just to make a point :-)

1. Note how I've added an [additional batch endpoint](https://github.com/creativesoftwarefdn/weaviate/blob/11e8677ca1633ae18744b97eedd30bc22a8d85fb/openapi-specs/schema.json#L1140) (@moretea a union is not possible because go-swagger doesn't support 3.0 specs).
2. There is a [separate function](https://github.com/creativesoftwarefdn/weaviate/blob/feature/example-GQL-batch/restapi/configure_weaviate.go#L1307) that handles the GQL stuff.
3. Note in [this function](https://github.com/creativesoftwarefdn/weaviate/blob/feature/example-GQL-batch/restapi/configure_weaviate.go#L1248) how there is a promise that handles every single request.

## Moral of the story

This might significantly simplify the implementation of batching

**Comment by moretea** [comment]
Sigh. @bobvanluijt this is _exactly_ what [I designed before](https://github.com/creativesoftwarefdn/weaviate/commit/daa9a9aebbe9fe80c02eff22326b8a8a7b700f8d).

However, you're [adding to a slice in goroutine](https://github.com/creativesoftwarefdn/weaviate/blob/feature/example-GQL-batch/restapi/configure_weaviate.go#L1272). This is _not_ thread safe. You might lose answers if you do this, if you're lucky.

**Comment by bobvanluijt** [comment]
@moretea, this has nothing to do with the swagger definition, but with the way it is handled in the configure-weaviate file. I’m failing to grasp what the difficulty is here.

I needed to update the swagger file to make my point and to validate how the `union` type functions in openapi specs 3.0.0.
